### PR TITLE
Abort inference requests when the HTTP client disconnects

### PR DESCRIPTION
**Pull request by santhnm2**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
A non-streaming completion writes nothing to the socket while it generates, so a peer that goes away mid-generation is never discovered as a broken pipe. The engine ran the request to its token limit while holding a batch slot. Under a workload that creates and destroys clients continuously the orphans accumulated faster than they retired: server-reported active requests grew monotonically with elapsed time, and once live plus orphaned requests exceeded max_num_seqs new requests queued behind work nobody was waiting for.

Quart already surfaces the disconnect. Its ASGI connection races handle_messages against handle_request with FIRST_COMPLETED and cancels the loser, and handle_messages returns on http.disconnect, so the handler coroutine receives CancelledError. Catch it, abort every request the handler submitted, and re-raise so the cancellation still propagates.

Aborting needs request ids, and add_request returned only a future -- cancelling that alone leaves the engine generating. Add add_request_with_id, which returns both; add_request now delegates to it and is unchanged for existing callers.

Also abort on the error path. asyncio.gather does not cancel siblings when one child raises, so a single failed request leaked its n-1 siblings the same way while the handler returned a 500.

The streaming path is untouched: it writes to the socket, so a dead peer surfaces as a broken pipe, and AsyncStream already carries an abort callback.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/common.py [hunk 1]
```diff
@@ -1,12 +1,39 @@
...
 
 import torch
 
 GENERATE_NUM = 0
 LOCK = threading.Lock()
 
+logger = logging.getLogger(__name__)
+
+
+def abort_requests(client, request_ids: Iterable[int], reason: str) -> None:
```
Comment (tdene) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/common.py]: nit: type `client` with a
```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from megatron.core.inference.inference_client import InferenceClient
```
and
`client: "InferenceClient"`

File: tests/unit_tests/inference/test_inference_client.py
Comment (tdene) [inline tests/unit_tests/inference/test_inference_client.py:1]: Please compress the new tests in this file to a single parameterized test. Parameterized tests are easier to read and maintain than individual, almost-identical, test methods.

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/chat_completions.py [hunk 2]
```diff
@@ -1043,17 +1044,47 @@ def parse_streaming_text(text):
...
+            abort_requests(client, request_ids, f"submission failed: {e}")
+            logger.error(f"Error submitting request: {e}")
+            return Response(f"Error submitting request: {e}", status=500)
 
         if current_app.config['verbose']:
             start_time = time.perf_counter()
 
         try:
             batch_results = await asyncio.gather(*tasks)
+        except asyncio.CancelledError:
```
Comment (tdene) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/chat_completions.py:1125]: There's currently no test for either of these `CancelledError` statements. Because of how finicky the intersections of these layers can be, I think it would be valuable to add a test that specifically uses Quart and exercises this path.

File: megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py [hunk 3]
```diff
@@ -214,7 +236,19 @@ async def completions():
...
+            # cancels the loser). Without this the engine would keep generating
+            # for a client that is gone, holding a slot until it hit the token
+            # limit -- orphans then accumulate faster than they retire and the
+            # batch saturates.
+            abort_requests(client, request_ids, "client disconnected")
+            raise
         except Exception as e:
+            # gather leaves siblings running when one child raises, so the
+            # others would leak the same way while we return a 500.
+            abort_requests(client, request_ids, f"inference error: {e}")
```
Comment (tdene) [inline megatron/core/inference/text_generation_server/dynamic_text_gen_server/endpoints/completions.py]: nit: we cannot currently hit the case where one child raises while sibling are running. The futures, currently, are left hanging unless explicitly resolved; they would have to be explicitly raised, and no one ever explicitly raises them. If a child request fails, the future never raises, it only completes with `status=="FAILED"`.

Given that, this `except Exception` cannot actually trigger in this codepath. I would personally remove it because it is confusing, and it may mask future issues if a mistake is made, and someone accidentally changes the design so that some inference errors successfully complete with `status=="FAILED"`, while other inference errors raise the future.

**Review by tdene — APPROVED** [review]
Looks good!

I left 3 nits, and 1 important comment. The Quart test request is what I consider important; I would like to see that test added before this is merged.

**Review by sidsingh-nvidia — APPROVED** [review]
Independently repro-ed the issue as well as the fix in swe-bench.

**Review by kvareddy — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]
